### PR TITLE
[Serve] Add metric to track `LongPollHost` transmissions

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

This change adds the `ray_serve_long_poll_host_transmission_counter` metric. This metric counts the total number of `LongPollHost` transmissions from the Serve controller.

## Related issue number

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change adds `test_long_poll_host_sends_counted` to test the new metric.

### CODE REVIEW

**Review by zcin — COMMENTED** [review]
Nice work!

File: doc/source/serve/monitoring.md [hunk 1]
```diff
@@ -456,6 +456,9 @@ The following metrics are exposed by Ray Serve:
      - * deployment
        * actor_id
      - The number of request scheduling tasks in the router that are undergoing backoff.
+   * - ``ray_serve_long_poll_host_bytes_sent`` [†]
```
Comment (edoakes) [inline doc/source/serve/monitoring.md:459]: I'd suggest leaving this one undocumented given that it's very much in the weeds of implementation details

File: python/ray/serve/_private/long_poll.py [hunk 2]
```diff
@@ -214,6 +223,22 @@ def _get_num_notifier_events(self, key: Optional[KeyType] = None):
         else:
             return sum(len(events) for events in self.notifier_events.values())
 
+    def count_bytes_returned(listen_for_change_func):
```
Comment (edoakes) [inline python/ray/serve/_private/long_poll.py:226]: add type hints for expected return value of `listen_for_change_func`

nit: it's also a bit of an abstraction leak to name it "listen_for_change_func"

**Review by edoakes — APPROVED** [review]
Just small nits

File: python/ray/serve/_private/long_poll.py [hunk 3]
```diff
@@ -193,6 +196,7 @@ def __init__(
         listen_for_change_request_timeout_s: Tuple[
             int, int
         ] = LISTEN_FOR_CHANGE_REQUEST_TIMEOUT_S,
+        long_poll_host_name_metrics_tag: str = "",
```
Comment (sihanwang41) [inline python/ray/serve/_private/long_poll.py:199]: nit: `long_poll_host_name`.

**Comment by sihanwang41** [comment]
Hi, this is not an accurate number even with measuring the non-wire size. Check
```
import sys
from typing import Any, NamedTuple
from dataclasses import dataclass

class DeploymentID(NamedTuple):
    name: str
    app: str

@dataclass
class UpdatedObject:
    object_snapshot: Any
    # The identifier for the object's version. There is not sequential relation
    # among different object's snapshot_ids.
    snapshot_id: int

a = UpdatedObject(1, {DeploymentID("1", "2"): 1})
print(sys.getsizeof(a))
b = UpdatedObject(1, {"1": 1})
print(sys.getsizeof(a))
```

It will print the same size. That is because `sys.getsizeof` only reports the memory to which this object is accounted for not including nested object inferring to. 

Two directions to go:
1. Push this metric into the core side. (if we can get the metrics per task request payload)
2. Write our own check size function instead of using `sys.getsizeof`.

File: python/ray/serve/_private/long_poll.py [hunk 4]
```diff
@@ -214,6 +223,22 @@ def _get_num_notifier_events(self, key: Optional[KeyType] = None):
...
+        number of bytes returned.
+        """
+
+        @wraps(listen_for_change_func)
+        async def count_bytes(self, *args, **kwargs):
+            objects = await listen_for_change_func(self, *args, **kwargs)
+            self.bytes_sent_counter.inc(value=sys.getsizeof(objects))
+            return objects
+
+        return count_bytes
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/long_poll.py:234]: Thanks for raising this, but due to [this issue](https://github.com/ray-project/ray/pull/39777#issuecomment-1731899162), I'll have to change the overall implementation.

File: python/ray/serve/_private/long_poll.py [hunk 5]
```diff
@@ -206,6 +210,11 @@ def __init__(
         )
 
         self._listen_for_change_request_timeout_s = listen_for_change_request_timeout_s
+        self.bytes_sent_counter = metrics.Counter(
+            "serve_long_poll_host_bytes_sent",
+            description="The number of bytes sent by the long poll host.",
+            tag_keys=("long_poll_host_name",),
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply python/ray/serve/_private/long_poll.py:211]: Yep, that's right.

File: doc/source/serve/monitoring.md [hunk 6]
```diff
@@ -456,6 +456,9 @@ The following metrics are exposed by Ray Serve:
...
+     - N/A
```
[the remark this answers is not in the document]
Comment (shrekris-anyscale) [reply doc/source/serve/monitoring.md:460]: Yeah, I don't think it'll be useful for users.

File: doc/source/serve/monitoring.md (hunk 1 above)
Comment (shrekris-anyscale) [reply doc/source/serve/monitoring.md:459]: Sure, I'm okay with that.

File: python/ray/serve/_private/long_poll.py (hunk 2 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/long_poll.py:221]: Thanks for raising this, but due to https://github.com/ray-project/ray/pull/39777#issuecomment-1731899162, I'll have to change the overall implementation.

File: python/ray/serve/_private/long_poll.py (hunk 3 above)
Comment (shrekris-anyscale) [reply python/ray/serve/_private/long_poll.py:194]: Thanks, renamed!

**Comment by shrekris-anyscale** [comment]
> That is because `sys.getsizeof` only reports the memory to which this object is accounted for not including nested object inferring to.

Thanks for the catch @sihanwang41! This is pretty surprising. I dug deeper into this, and I don't think there's a safe or straightforward way to calculate the true size of an object in Python. I'll have to change the overall implementation of this metric.

My plan is to track the number of broadcasts per namespace instead. It's not ideal, but it'll at least give us an understanding of when the `LongPollHost` is sending traffic.

**Comment by shrekris-anyscale** [comment]
@zcin @edoakes @sihanwang41 I updated this PR with a new implementation. Instead of counting bytes, the metric now counts total transmissions. The metric is tagged by namespace and state, so we know what the objects are being sent. Please take another look.

**Review by sihanwang41 — APPROVED** [review]

**Pull request merged**